lib/jekyll/site.rb

```diff
@@ -326,14 +326,14 @@ def relative_permalinks_are_deprecated
     #
     # Returns an Array of Documents which should be written
     def docs_to_write
-      @docs_to_write ||= documents.select(&:write?)
+      documents.select(&:write?)
     end
 
     # Get all the documents
     #
     # Returns an Array of all Documents
     def documents
-      @documents ||= collections.reduce(Set.new) do |docs, (_, collection)|
+      collections.reduce(Set.new) do |docs, (_, collection)|
```

Comment:
Can you retest the memory allocations for each_with_object instead of reduce? The order of the block arguments are slightly different and you’d do `docs.merge(collection.docs)`.

The idea here is that the reduce method allocations a number of object in each iteration, but each_with_object simply iterates over your existing objects and lets you build the set with each iteration.